text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+
+## Mirroring
+
+There is currently no way for end users to cryptographically guarantee, in isolation, that an index entry, crate package, or Rust release was definitively sourced from the Rust Project. Although the entirety of the current end-to-end system provides some protections (hashes in the index of crate files, and HTTPS for all communications), this precludes many out-of-band mirroring scenarios and is subject to multiple attack vectors.
+
+This proposal adds cryptographic security for multiple mirroring scenarios:
+- Rust releases (downloaded directly or via `rustup`)
+- `crates.io` index
+- `crates.io` packages
+
+Today, for `crates.io` and official infrastructure, we strictly distribute from Rust Project controlled sources (our CDN network). This prevents us from leveraging mirroring offers from third parties, such as academic institutions, to host copies of crates.io packages as well as Rust releases. Having a set of root cryptographic keys gives us the tools to begin exploring multiple methods of securing both internal and external mirroring and propagation problems.
+
+We leave the implementation of these solutions for a future RFC - however, this provides us with the first steps in having the available cryptographic keys to begin exploring and discussing solutions to these problems.
+
+## Private Repositories
+
+The technical details of this proposal involve the initial signing of the index and crate files utilizing the Rust Project root key for validation. However, the implementation of this in cargo will allow for private repositories to validate against additional certificates.  This means all new security guarantees made with the new Rust keys can also be leveraged by private repositories for private certificate chains. (Cargo will not, however, allow the substitution of different keys for `crates.io` itself.)
+
+## Provenance
+
+Utilizing our PKI and certificate hierarchy, end users (manually or utilizing `rustup` and `cargo`) will be able to cryptographically verify the provenance of arbitrary files, regardless of source, as being from the Rust Project.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We propose leveraging [AWS CloudHSM Quorum Authentication](https://docs.aws.amazon.com/cloudhsm/latest/userguide/quorum-authentication.html) for the management of the `Root Key`. See [Alternatives#Quorum] for out-of-band quorum models which were considered.
+
+All operations using the root key or subkeys will be logged transparently, and will make use of Certificate Transparency / Binary Transparency infrastructure to ensure that no surreptitious signing operations can take place.
+
+## Keys and Trust
+
+**Key Algorithm:** ECDSA secp384r1
```

Comment:
Why ECDSA and not EdDSA? From https://www.rfc-editor.org/rfc/rfc8032#section-1:

> The advantages with EdDSA are as follows:
>
>    1.  EdDSA provides high performance on a variety of platforms;
>
>    2.  The use of a unique random number for each signature is not
>        required;

ECDSA fails catastrophically with nonce reuse. The PS3 was broken this way for example. EdDSA doesn't require any randomness for producing signatures at all.

>
>    3.  It is more resilient to side-channel attacks;
>
>    4.  EdDSA uses small public keys (32 or 57 bytes) and signatures (64
>        or 114 bytes) for Ed25519 and Ed448, respectively;
>
>    5.  The formulas are "complete", i.e., they are valid for all points
>        on the curve, with no exceptions.  This obviates the need for
>        EdDSA to perform expensive point validation on untrusted public
>        values; and
>
>    6.  EdDSA provides collision resilience, meaning that hash-function
>        collisions do not break this system (only holds for PureEdDSA).

Reply:
The certificates in the PKI introduced by this PR wouldn't be trusted by browsers, right? So it doesn't matter if browsers can't parse the certificates. That said: Rustls has ed25519 support (both the ring and aws_lc_rs crypto backends) and a fixme for ed448 support: https://github.com/rustls/rustls/blob/cf098b08d32b8f9636faf3dd82a1b8d4a85a146e/rustls/src/crypto/ring/sign.rs#L71-L75 OpenSSL supports ed25519 and ed448. BoringSSL supports ed25519 but not ed448. Same for firefox's nss. In other words as far as I can tell ed25519 would meet your criteria. If the security margin of ed448 over ed25519 is really important, shouldn't a hybrid scheme with a post quantum signature algorithm be used?

Reply:
> The other question, of course, is what CloudHSM supports. That, unfortunately, is a much more limited list:

Right, ... What about giving the root a sep384r1 key under the assumption that the HSM is secure against timing side channel attacks (by the virtue of being offline most of the time and only being operated by trusted people) and implementation bugs (surely cloudflare has audited it, right?) and use eddsa for all certificates signed with the root certificate as these aren't protected by an HSM?

Reply:
We're about to push an update that specifically states that delegate keys can use different algorithms.